### DISCUSSION TITLE
How to access `lua.used_memory()` from `lua.set_interrupt()` callback?

### DISCUSSION THREAD

**Discussion by 33KK** [mlua-rs/mlua#196] [Q&A]
I need to cap memory usage of the lua context, so I am trying to check current memory usage in interrupt handler since `set_memory_limit` is not available with Luau. Using an `Arc<Mutex<Lua>>` doesn't work either since it doesn't drop the lock while executing code.

EDIT: also i'm not quite sure what is the right way to stop execution, `return Err(mlua::Error::RuntimeError("script took too long to execute".to_owned()));` seems to work but doesn't feel right

**Comment by 33KK** [comment]
It seems that `set_memory_limit` works fine with luau, any reason it is not enabled for `luau` feature?

**Comment by khvzak** [comment]
> any reason it is not enabled for luau feature?

Same reason as https://github.com/khvzak/mlua/issues/119

> Using an Arc<Mutex<Lua>> doesn't work either since it doesn't drop the lock while executing code.

You probably don't need `Mutex` here, and just can try `Arc`/`Rc` instead... Unfortunately Luau does not provide fully defined semantics for interrupts (but checking memory should be ok).

>  also i'm not quite sure what is the right way to stop execution, return Err(mlua::Error::RuntimeError("script took too long to execute".to_owned())); seems to work but doesn't feel right

More elegant way is to trigger `VmState::Yield` to force return execution back to Rust. Generating an exception is also fine.
An example: https://github.com/khvzak/mlua/blob/v0.8.3/tests/luau.rs#L176

**Reply by 33KK** [reply]
![image](https://user-images.githubusercontent.com/23123008/182485867-95e3375a-0987-476e-80be-1dab3131f284.png)
Using an `Arc` doesn't seem to work, I do have the send feature enabled

**Reply by khvzak** [reply]
Seems the only way is to use [`ReentrantMutex`](https://docs.rs/parking_lot/0.12.1/parking_lot/type.ReentrantMutex.html) from parking_lot :(
I see how inconvenient it is! I probably change interrupts interface in a next major release.

**Reply by 33KK** [reply]
I see, I guess I'll convert this into an issue for now
